purchase_product_pack/models/purchase_order_line.py

```diff
@@ -0,0 +1,211 @@
+                else:
+                    vals_list.append(vals)
+            if vals_list:
+                self.create(vals_list)
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        new_vals = []
+        res = self.browse()
+        for elem in vals_list:
+            product = self.env["product.product"].browse(elem.get("product_id"))
+            if product and product.pack_ok and product.pack_type != "non_detailed":
+                line = super().create([elem])
+                line.expand_pack_line()
+                res |= line
+            else:
+                new_vals.append(elem)
+        res |= super().create(new_vals)
+        return res
+
+    def write(self, vals):
+        res = super().write(vals)
+        if "product_id" in vals or "product_qty" in vals:
+            for record in self:
+                record.expand_pack_line(write=True)
+        return res
+
+    @api.onchange(
+        "product_id",
+        "product_uom_qty",
+        "product_uom",
+        "price_unit",
+        "name",
+        "taxes_id",
+    )
+    def check_pack_line_modify(self):
+        """Do not let to edit a purchase order line if this one belongs to pack"""
+        if self._origin.pack_parent_line_id and not self._origin.pack_modifiable:
+            raise UserError(
+                _(
+                    "You can not change this line because is part of a pack"
+                    " included in this order"
+                )
+            )
+
+    def action_open_parent_pack_product_view(self):
+        domain = [
+            ("id", "in", self.mapped("pack_parent_line_id").mapped("product_id").ids)
+        ]
+        return {
+            "name": _("Parent Product"),
+            "type": "ir.actions.act_window",
+            "res_model": "product.product",
+            "view_type": "form",
+            "view_mode": "tree,form",
+            "domain": domain,
+        }
+
+    @api.depends("product_qty", "product_uom")
+    def _compute_price_unit_and_date_planned_and_name(self):
+        res = super()._compute_price_unit_and_date_planned_and_name()
+        # Re-calculate the price unit following options on product-pack
+        for line in self:
+            if not line.product_id or line.invoice_lines:
+                continue
+
+            params = {"order_id": line.order_id}
+            seller = line.product_id._select_seller(
+                partner_id=line.partner_id,
+                quantity=line.product_qty,
+                date=line.order_id.date_order and line.order_id.date_order.date(),
+                uom_id=line.product_uom,
+                params=params,
+            )
+
+            prices = line.product_id.with_context(line=line).cost_compute()
+            if not prices:
+                continue
+            else:
+                cost = prices[line.product_id.id]
+            # If not seller, use the standard price. It needs a proper currency conversion.
+            if not seller:
```

Comment:
this block should be moved out to its own method

Reply:
hmm I see this method has been partially copied from the odoo core one. Can you mention this and highlight the parts that have been modified?